### ISSUE TITLE
Empty action-parameter vector is always accessed in ActionFnEntry::execute

### ISSUE THREAD

**Issue by kjw6855** [p4lang/behavioral-model#1230]
### Environment
- ONOS v2.7.1
- Stratum v2022-06-30
- BMv2 b0fb01e (but main branch has the same code)
- Stratum and BMv2 are built with LLVM_UBSAN and LLVM_ASAN

### Description
After connecting to ONOS, out-of-bound memory access occurs.
In `ActionFnEntry::execute(Packet *pkt)`, `action_fn->params` can be empty.
The causal action's primitive seems to have a parsing error.
But, the code always access action_fn->params.
```
    param_offset = primitive.get_param_offset();
    primitive.execute(&state, &(action_fn->params[param_offset]));
```
Suggested solution:
```
    param_offset = primitive.get_param_offset();
    primitive.execute(&state, (action_fn->params.size() > param_offset) ?
            &(action_fn->params[param_offset]) : nullptr);
```

**Comment by jafingerhut** [comment]
Are you compiling a particular P4 program to product a BMv2 JSON "executable" file that is being loaded into the `simple_switch` or `simple_switch_grpc` process, and that process is crashing?

If you can find a way to reproduce this issue outside of the context of running ONOS (or can find someone who can help you to create such a reproduction), that would be easier for others to diagnose what is going on.

**Comment by kjw6855** [comment]
It doesn't happen while installing P4 programs. Also, I saw this error in several P4 programs: e.g, basic.p4 and fabric.p4.
I don't think it's easy to reproduce without ONOS.

FYI, let me share lldb output when it occurs:
```
* thread #6, name = 'stratum_bmv2', stop reason = signal SIGILL: illegal instruction operand
    frame #0: 0x00007ffff662556d libsimpleswitch_runner.so.0`std::vector<bm::ActionParam, std::allocator<bm::ActionParam> >::operator[](this=size=1, __n=0) const at stl_vector.h:1061:2
   1058	      operator[](size_type __n) const _GLIBCXX_NOEXCEPT
   1059	      {
   1060		__glibcxx_requires_subscript(__n);
-> 1061		return *(this->_M_impl._M_start + __n);
   1062	      }
   1063
   1064	    protected:

* thread #6, name = 'stratum_bmv2', stop reason = signal SIGILL: illegal instruction operand
    frame #0: 0x00007ffff662556d libsimpleswitch_runner.so.0`std::vector<bm::ActionParam, std::allocator<bm::ActionParam> >::operator[](this=size=1, __n=0) const at stl_vector.h:1061:2
  * frame #1: 0x00007ffff66244cb libsimpleswitch_runner.so.0`bm::ActionFnEntry::execute(this=0x000061600003aa88, pkt=0x000060f000007a20) const at actions.cpp:411:33
    frame #2: 0x00007ffff662709d libsimpleswitch_runner.so.0`bm::ActionFnEntry::operator(this=0x000061600003aa88, pkt=0x000060f000007a20)(bm::Packet*) const at actions.cpp:429:5
    frame #3: 0x00007ffff692f296 libsimpleswitch_runner.so.0`bm::MatchTableAbstract::apply_action(this=0x000061600003a880, pkt=0x000060f000007a20) at match_tables.cpp:138:3
    frame #4: 0x00007ffff6d25982 libsimpleswitch_runner.so.0`bm::MatchActionTable::operator(this=0x000060600009d580, pkt=0x000060f000007a20)(bm::Packet*) const at tables.cpp:42:34
    frame #5: 0x00007ffff6c2f055 libsimpleswitch_runner.so.0`bm::Pipeline::apply(this=0x00006060000a2b00, pkt=0x000060f000007a20) at pipeline.cpp:44:12
    frame #6: 0x00007ffff627f8a3 libsimpleswitch_runner.so.0`SimpleSwitch::egress_thread(this=0x000061a000000080, worker_id=1) at simple_switch.cpp:677:17
    frame #7: 0x00007ffff628f6d5 libsimpleswitch_runner.so.0`void std::__invoke_impl<void, void (SimpleSwitch::*)(unsigned long), SimpleSwitch*, unsigned long>((null)=__invoke_memfun_deref @ 0x00007fffefdf9920, __f=0x000060400003d828, __t=0x000060400003d820, __args=0x000060400003d818)(unsigned long), SimpleSwitch*&&, unsigned long&&) at invoke.h:73:14
    frame #8: 0x00007ffff628f3ab libsimpleswitch_runner.so.0`std::__invoke_result<void (SimpleSwitch::*)(unsigned long), SimpleSwitch*, unsigned long>::type std::__invoke<void (__fn=0x000060400003d828, __args=0x000060400003d820, __args=0x000060400003d818)(unsigned long), SimpleSwitch*, unsigned long>(void (SimpleSwitch::*&&)(unsigned long), SimpleSwitch*&&, unsigned long&&) at invoke.h:95:14
    frame #9: 0x00007ffff628eda9 libsimpleswitch_runner.so.0`void std::thread::_Invoker<std::tuple<void (SimpleSwitch::*)(unsigned long), SimpleSwitch*, unsigned long> >::_M_invoke<0ul, 1ul, 2ul>(this=0x000060400003d818, (null)=_Index_tuple<0UL, 1UL, 2UL> @ 0x00007fffefdf99a8) at thread:244:13
    frame #10: 0x00007ffff628ea6d libsimpleswitch_runner.so.0`std::thread::_Invoker<std::tuple<void (SimpleSwitch::*)(unsigned long), SimpleSwitch*, unsigned long> >::operator(this=0x000060400003d818)() at thread:251:11
    frame #11: 0x00007ffff628e9e5 libsimpleswitch_runner.so.0`std::thread::_State_impl<std::thread::_Invoker<std::tuple<void (SimpleSwitch::*)(unsigned long), SimpleSwitch*, unsigned long> > >::_M_run(this=0x000060400003d810) at thread:195:13
    frame #12: 0x00007ffff5555df4 libstdc++.so.6`___lldb_unnamed_symbol7640 + 20
    frame #13: 0x00007ffff57ba609 libpthread.so.0`start_thread(arg=<unavailable>) at pthread_create.c:477:8
    frame #14: 0x00007ffff5391353 libc.so.6`__clone at clone.S:95
```

**Comment by jafingerhut** [comment]
I do not know how many people are reading issues created for the behavioral-model that might know about this, but hopefully someone will respond.

I would recommend checking in any ONOS help or support forums you know of, to see if anyone there has run across this behavior.

**Comment by kjw6855** [comment]
Here are some updated information.
Please take a look at it.

### New environment
- `simple_switch` @ BMv2 11c789b (Aug 7th, 2024).
- `simple_switch` is built with clang-14 LLVM_UBSAN and LLVM_ASAN.
(Note: the compilation has failed with targets. I recommend to compile `simple_switch` separately as follows.)
```
$ CC=clang CXX=clang++ LD=clang ./configure --without-targets --with-pi --disable-elogger --without-nanomsg --with-thrift --prefix=/usr/local CXXFLAGS="-I$(pwd)/targets/simple_switch -DWITH_SIMPLE_SWITCH -isystem/include -isystem/usr/local/include -L/usr/local/lib -fsanitize=address -fsanitize=undefined" LDFLAGS="-fsanitize=address -fsanitize=undefined"
$ make && sudo make install && sudo ldconfig
$ cd targets/simple_switch
$ make && sudo make install && sudo ldconfig

# (a workaround to link simple_switch binary)
$ /bin/bash ../../libtool  --tag=CXX  --mode=link clang++ -Wall -Wextra -pthread -O0 -g \
-fsanitize=address -fsanitize=undefined -L/usr/local/lib -I$(pwd) \
-isystem/include -isystem/usr/local/include -rdynamic -o simple_switch \
main.o thrift/src/.libs/SimpleSwitch_server.o ../../src/.libs/libbmall.la \
../../thrift_src/.libs/libruntimestubs.la libsimpleswitch.la libsimpleswitch_thrift.la \
libsimpleswitch_runner.la -lboost_thread -latomic -lpcap -lgmp -lnanomsg -lthrift
```

<details>
  <summary> Sample P4 program (issue_1230.p4) </summary>

  ```
  // Modified from pvs_struct_2.p4
  #include <v1model.p4>
  
  header data_h {
    bit<32> da;
    bit<32> db;
  }
  
  struct my_packet {
    data_h data;
  }
  
  struct my_metadata {
    data_h[2] data;
  }
  
  parser MyParser(packet_in b, out my_packet p, inout my_metadata m, inout standard_metadata_t s) {
      state start {
          transition accept;
      }
  }
  
  control MyVerifyChecksum(inout my_packet hdr, inout my_metadata meta) {
    apply { }
  }
  
  
  control MyIngress(inout my_packet p, inout my_metadata meta, inout standard_metadata_t s) {
      apply {
          exit;  // Erroneous statement
      }
  }
  
  control MyEgress(inout my_packet p, inout my_metadata m, inout standard_metadata_t s) {
    apply {
    }
  }
  
  control MyComputeChecksum(inout my_packet p, inout my_metadata m) {
    apply { }
  }
  
  control MyDeparser(packet_out b, in my_packet p) {
    apply { }
  }
  
  V1Switch(MyParser(), MyVerifyChecksum(), MyIngress(), MyEgress(), MyComputeChecksum(), MyDeparser()) main;
  ```

</details>


### Reproducing steps
1. Setup veth
`$ ./tools/veth_setup.sh`
2. Run simple_switch with the issue_1230.p4 program
(You might need to configure `ASAN_OPTIONS=detect_odr_violation=0`.)
`$ ASAN_OPTIONS=detect_odr_violation=0 sudo simple_switch -i 1@veth0 -i 2@veth2 -i 3@veth4 issue_1230.json --log-console`
3. Send ping to any port (e.g., `veth5`)
```
from scapy.all import *
packet = Ether()/IP()/ICMP()
sendp(packet, iface='veth5')
```
4. Following error will pop up
```
[17:52:39.072] [bmv2] [T] [thread 134564] [0.0] [cxt 0] issue_1230.p4(32) Primitive exit
/usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_vector.h:1061:9: runtime error: reference binding to null pointer of type 'const __gnu_cxx::__alloc_traits<std::allocator<bm::ActionParam>, bm::ActionParam>::value_type' (aka 'const bm::ActionParam')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/stl_vector.h:1061:9 in
```

### Root cause
In `issue_1230.p4`, there is an `exit` statement.
The block with `exit` is translated as a `exit` primitive without any parameter.
```
"actions" : [
  {
    "name" : "issue_1230l32",
    "id" : 0,
    "runtime_data" : [],
    "primitives" : [
      {
        "op" : "exit",
        "parameters" : [],
        "source_info" : {
          "filename" : "issue_1230.p4",
          "line" : 32,
          "column" : 8,
          "source_fragment" : "exit"
        }
      }
    ]
  }
],
```

### Recommended patch
`ActionFnEntry::execute(Packet *pkt)` in `src/bm_sim/actions.cpp` should check whether `param_offset` is in the range of `action_fn->params`.
Here is the suggested patch: 
```
diff --git a/src/bm_sim/actions.cpp b/src/bm_sim/actions.cpp
index 24b76cc..a7bed29 100644
--- a/src/bm_sim/actions.cpp
+++ b/src/bm_sim/actions.cpp
@@ -408,7 +408,10 @@ ActionFnEntry::execute(Packet *pkt) const {
         (primitive.get_source_info() == nullptr) ? "(no source info)"
         : primitive.get_source_info()->get_source_fragment());
     param_offset = primitive.get_param_offset();
-    primitive.execute(&state, &(action_fn->params[param_offset]));
+    // action_fn (string parsing error) can have no params
+    if (action_fn->params.size() > param_offset) {
+        primitive.execute(&state, &(action_fn->params[param_offset]));
+    }
     idx = primitive.get_jump_offset(idx);
   }
 }
```

### Relationship with ONOS's [basic.p4](https://gerrit.onosproject.org/plugins/gitiles/onos/+/refs/heads/master/pipelines/basic/src/main/resources/basic.p4)
In `include/packet_io.p4` of basic.p4, there is an `exit` statement.
```
# include/packet_io.p4
control packetio_ingress(inout headers_t hdr,
                         inout standard_metadata_t standard_metadata) {
    apply {
        if (standard_metadata.ingress_port == CPU_PORT) {
            standard_metadata.egress_spec = hdr.packet_out.egress_port;
            hdr.packet_out.setInvalid();
            exit;
        }
    }
}
```
This `exit` in the if block is translated as one of primitives in `act_1` without any parameter.
```
# basic.json
"name" : "act_1",
"id" : 10,
"runtime_data" : [],
"primitives" : [
...
{
  "op" : "exit",
  "parameters" : [],
  "source_info" : {
    "filename" : "include/packet_io.p4",
    "line" : 29,
    "column" : 12,
    "source_fragment" : "exit"
  }
}
],
```
To test the error with basic.p4, you can run following commands. However, in my environment UBSAN doesn't catch the out-of-bound access in `simple_switch` binary. Instead, I confirmed the out-of-bound access with a debugger (e.g., lldb). Here is how to reproduce it:
1. Run `simple_switch` with a debugger. Allocate one of veth interfaces as 255 (CPU_PORT). Create a breakpoint at `actions.cpp:427`:
```
$ sudo lldb simple_switch
(lldb) r -i 1@veth0 -i 2@veth2 -i 255@veth4 $HOME/onos/pipelines/basic/src/main/resources/p4c-out/bmv2/basic.json` --log-console
^C
(lldb) b actions.cpp:427
(lldb) c
```
3. Send ping to port 255 (i.e., `veth5`)
4. Out-of-bound read in `action_fn->param` occurs.
```
[07:27:26.815] [bmv2] [T] [thread 99694] [1.0] [cxt 0] include/packet_io.p4(29) Primitive exit

- Hook 1 (expr -- action_fn->params.size())
(size_type) $11 = 3

- Hook 2 (expr -- param_offset)
(size_t) $12 = 3
Process 99686 stopped
* thread #7, name = 'simple_switch', stop reason = breakpoint 2.1
    frame #0: 0x0000555555dc06d1 simple_switch`bm::ActionFnEntry::execute(this=0x0000616000006b88, pkt=0x000060f000007930) const at actions.cpp:427:5
   424 	        (primitive.get_source_info() == nullptr) ? "(no source info)"
   425 	        : primitive.get_source_info()->get_source_fragment());
   426 	    param_offset = primitive.get_param_offset();
-> 427 	    primitive.execute(&state, &(action_fn->params[param_offset]));
   428 	    idx = primitive.get_jump_offset(idx);
   429 	  }
   430 	}
```

**Comment by jafingerhut** [comment]
Thanks for the detailed reproduction steps, especially ones that do not require lots of other software to be installed to trigger it!

**Comment by antoninbas** [comment]
Thanks for the detailed report.
The patch is not valid, as `primitive.execute` still needs to be called in that case.
While it may be *technically* undefined behavior, this will never be a problem in practice as the pointer obtained this way will never be dereferenced.
When your if condition evaluates to false, you could use null as the argument:
```cpp
    if (action_fn->params.size() > param_offset) {
        primitive.execute(&state, &(action_fn->params[param_offset]));
    } else {
        primitive.execute(&state, nullptr));
    }
```
Personally I prefer avoiding unnecessary if stetements, and this should avoid the asan error:
```cpp
    primitive.execute(&state, action_fn->params.data() + param_offset);
```
In this case we get the desired value using pointer arithmetic, so there is should be no UB

**Comment by kjw6855** [comment]
You're right! I forgot to add `primitive.execute()` in actions without parameters.

I tested your code with UBSAN, there is no more a sanitization error in my environment.
However, the `param_offset` could be the same as `params.size()`, which is still (technically) undefined behavior.
For example, param_offset was 3 with a three-length vector in ONOS `basic.p4` since the block contains multiple actions with exit.
In my humble opinion, following statement looks good, but the decision depends on P4 developers.
```
primitive.execute(&state, action_fn->params.size() > param_offset ? &(action_fn->params[param_offset]) : nullptr);
```

Also, it seems that ActionPrimitives (e.g. exit_()) can only access the predefined number of arguments (e.g., no args for exit).
I directly modified the issue_1230.json (output of P4C) as well. But BMv2 doesn't allow malformed operations (e.g., assign without parameters) at all.

Thank you for all responses.

**Comment by antoninbas** [comment]
I don't see how `action_fn->params.data() + param_offset` is UB even when `param_offset == action_fn->params.size()`. It computes a pointer that will never be dereferenced. It's different from `&action_fn->params[param_offset]` which I agree would technically be UB (at least in modern C++, I am not sure is UB in C as long as you don't dereference the resulting pointer). I prefer to avoid conditionals when they are not truly needed.

> Also, it seems that ActionPrimitives (e.g. exit_()) can only access the predefined number of arguments (e.g., no args for exit).
I directly modified the issue_1230.json (output of P4C) as well. But BMv2 doesn't allow malformed operations (e.g., assign without parameters) at all.

I am not sure what the point is here? Yes, I imagine you would get errors if the number of arguments provided doesn't match the length of the parameter list for the primitive. IIRC, there should be an error when bmv2 first loads the JSON.

**Comment by kjw6855** [comment]
If I understand it correctly, while the argument value (i.e. `action_fn->params.data() + param_offset`) is calculated and it points out of range of param vector, the argument pointer is never dereferenced by the following code, so we can ignore it.

> I directly modified the issue_1230.json (output of P4C) as well. But BMv2 doesn't allow malformed operations (e.g., assign without parameters) at all.

Regarding this, your explanation is correct. I cannot run BMv2 with malformed JSON and it shows an error.
I tested BMv2 in this way, since I found that BMv2 tries to support previous versions of P4 (e.g., `drop` which doesn't have any param and can be called within the block, instead of `mark_to_drop`).
